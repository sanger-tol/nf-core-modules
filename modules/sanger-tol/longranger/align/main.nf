process LONGRANGER_ALIGN {
    tag "${meta.id}"
    label 'process_high'

    // NOTE: As longranger is proprietary software, no container or conda environment
    // can be provided for this module. Please configure your own container and then
    // pass the path to this container as part of a Nextflow configuration file.

    input:
    tuple val(meta) , path(fastq, stageAs: "10X_inputs/*")
    tuple val(meta2), path(reference)

    output:
    tuple val(meta), path("${prefix}/outs/possorted_bam.bam")    , emit: bam
    tuple val(meta), path("${prefix}/outs/possorted_bam.bam.bai"), emit: bai
    tuple val(meta), path("${prefix}/outs/summary.csv")          , emit: csv
    path("versions.yml")                                         , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def local_setup = !(task.ext.args =~ /--jobmode=lsf|--jobmode=sge/) ? "--localcores=${task.cpus}" : ""
    def args        = task.ext.args   ?: ""
    prefix      = task.ext.prefix ?: "${meta.id}"
    """
    longranger align \\
        --id=${prefix} \\
        --fastqs=10X_inputs \\
        --sample=${prefix} \\
        --reference=${reference} \\
        ${local_setup} \\
        ${args}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        longranger: \$(longranger align --version | grep longranger | sed 's/.*(//' | sed 's/).*//')
    END_VERSIONS
    """

    stub:
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    mkdir -p ${prefix}/outs
    touch ${prefix}/outs/possorted_bam.bam
    touch ${prefix}/outs/possorted_bam.bam.bai
    touch ${prefix}/outs/summary.csv

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        longranger: \$(longranger align --version | grep longranger | sed 's/.*(//' | sed 's/).*//')
    END_VERSIONS
    """
}
