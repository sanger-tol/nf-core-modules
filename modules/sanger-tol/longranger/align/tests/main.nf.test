nextflow_process {

    name "Test Process LONGRANGER_ALIGN"
    script "../main.nf"
    process "LONGRANGER_ALIGN"
    config "./nextflow.config"

    tag "modules"
    tag "modules_sangertol"
    tag "longranger"
    tag "longranger/mkref"
    tag "longranger/align"

    setup {
        nfcoreInitialise("${launchDir}/library/")
        nfcoreInstall("${launchDir}/library/", ["gunzip"])
        nfcoreLink("${launchDir}/library/", "${baseDir}/modules")

        run("GUNZIP"){
            script "../../../../nf-core/gunzip/main.nf"
            process {
                """
                input[0] = [
                    [ id: "baUndUnlc1" ],
                    file(params.modules_testdata_base_path + 'Undibacterium_unclassified/assembly/draft/baUndUnlc1.hic.hap1.p_ctg.fa.gz', checkIfExists: true)
                ]
                """
            }
        }

        run("LONGRANGER_MKREF") {
            script "../../mkref/main.nf"
            process {
                """
                input[0] = GUNZIP.out.gunzip
                """
            }
        }
    }

    test("undibacterium unclassified") {

        when {

		    params {
                longranger_align_args = "--disable-ui --nopreflight"
		    }

			process {
                """
                input[0] = [
                    [ id: "baUndUnlc1" ],
                    [
						file(params.modules_testdata_base_path + "Undibacterium_unclassified/genomic_data/baUndUnlc1/10x/baUndUnlc1_S12_L002_R1_001.fastq.gz"),
						file(params.modules_testdata_base_path + "Undibacterium_unclassified/genomic_data/baUndUnlc1/10x/baUndUnlc1_S12_L002_R2_001.fastq.gz"),
						file(params.modules_testdata_base_path + "Undibacterium_unclassified/genomic_data/baUndUnlc1/10x/baUndUnlc1_S12_L002_I1_001.fastq.gz")
					]
                ]
                input[1] = LONGRANGER_MKREF.out.reference
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    bam(process.out.bam.get(0).get(1)).getStatistics(),
                    file(process.out.bai.get(0).get(1)).getName(),
                    process.out.csv,
                    process.out.versions
                ).match() }
            )
        }

        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules")
        }

    }

    test("undibacterium unclassified - stub") {

        options "-stub"

        when {

		    params {
                longranger_align_args = "--disable-ui --nopreflight"
		    }

            process {
                """
                input[0] = [
                    [ id: "baUndUncl1" ],
					[
						file(params.modules_testdata_base_path + "Undibacterium_unclassified/genomic_data/baUndUnlc1/10x/baUndUnlc1_S12_L002_R1_001.fastq.gz"),
						file(params.modules_testdata_base_path + "Undibacterium_unclassified/genomic_data/baUndUnlc1/10x/baUndUnlc1_S12_L002_R2_001.fastq.gz"),
						file(params.modules_testdata_base_path + "Undibacterium_unclassified/genomic_data/baUndUnlc1/10x/baUndUnlc1_S12_L002_I1_001.fastq.gz")
					]
                ]
                input[1] = LONGRANGER_MKREF.out.reference
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules")
        }

    }

}
