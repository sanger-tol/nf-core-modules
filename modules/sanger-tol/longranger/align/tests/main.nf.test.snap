{
    "undibacterium unclassified": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "bai": [
                    
                ],
                "bam": [
                    
                ],
                "csv": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-03T16:17:23.279938007"
    },
    "undibacterium unclassified - stub": {
        "content": [
            {
                "0": [
                    
                ],
                "1": [
                    
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "bai": [
                    
                ],
                "bam": [
                    
                ],
                "csv": [
                    
                ],
                "versions": [
                    
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-03T16:17:55.983594319"
    }
}